### PR TITLE
feat: support custom writer and storage for apply and preview

### PR DESCRIPTION
**Pull request by elliotxx**
Purpose:
* Public apply and preview func for external system integration
* Support custom state storage and writer for apply and preview func
* Bump the `github.com/pterm/pterm` package to the latest version

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -41,7 +41,7 @@ require (
 	github.com/mitchellh/go-wordwrap v1.0.1 // indirect
 	github.com/mitchellh/hashstructure v1.0.0
 	github.com/pkg/errors v0.9.1
-	github.com/pterm/pterm v0.12.41
+	github.com/pterm/pterm v0.12.42-0.20220427210824-6bb8c6e6cc77
```
[the remark this answers is not in the document]
Comment (elliotxx) [reply go.mod:44]: The PR use unrelease features of `pterm`

**Review by howieyuen — APPROVED** [review]

**Review by howieyuen — APPROVED** [review]

File: pkg/kusionctl/cmd/apply/options.go [hunk 2]
```diff
@@ -129,8 +131,14 @@ func (o *ApplyOptions) Run() error {
 	return nil
 }
 
-func preview(o *ApplyOptions, planResources *models.Spec,
-	project *projectstack.Project, stack *projectstack.Stack,
+// The preview function will preview for all resources changes
+func Preview(
+	o *ApplyOptions,
+	storage states.StateStorage,
```
Comment (ldxdl) [inline pkg/kusionctl/cmd/apply/options.go:137]: argument not used

File: pkg/kusionctl/cmd/apply/options.go [hunk 3]
```diff
@@ -129,8 +131,14 @@ func (o *ApplyOptions) Run() error {
 	return nil
 }
 
-func preview(o *ApplyOptions, planResources *models.Spec,
-	project *projectstack.Project, stack *projectstack.Stack,
+// The preview function will preview for all resources changes
```
Comment (ldxdl) [inline pkg/kusionctl/cmd/apply/options.go:134]: Write full docstring for public apis. See
https://github.com/KusionStack/KCLVM/blob/eb32a3d87eac48450fa4fb6331470cb735c1cf44/kclvm/tools/src/query/mod.rs#L21
https://github.com/KusionStack/KCLVM/blob/eb32a3d87eac48450fa4fb6331470cb735c1cf44/kclvm/tools/src/query/mod.rs#L21

File: pkg/kusionctl/cmd/apply/options.go [hunk 4]
```diff
@@ -166,7 +174,15 @@ func preview(o *ApplyOptions, planResources *models.Spec,
 	return operation.NewChanges(project, stack, rsp.Order), nil
 }
 
-func apply(o *ApplyOptions, planResources *models.Spec, changes *operation.Changes) error {
+// The apply function will apply the resources changes,
+// and will save the state to specified storage.
+func Apply(
```
Comment (ldxdl) [inline pkg/kusionctl/cmd/apply/options.go:179]: Check your pub apis, you'd better not write too many details in the pub api directly, try to keep the core logic/flow clear and readable, and preferably no more than one screen in length.

File: pkg/kusionctl/cmd/apply/options.go (hunk 2 above)
Comment (elliotxx) [reply pkg/kusionctl/cmd/apply/options.go:164]: nice catch, fixed

**Review by healthjyk — COMMENTED** [review]
look good to me

**Review by howieyuen — APPROVED** [review]

**Comment by ldxdl** [comment]
The interface design of the engine and the options’ pub API is not very reasonable. As here, the runtime interface shouldn't be exposed. @elliotxx and @SparkYuan, pls pay more attention to design and work out a more reasonable API interfaces of option and engine layer. And, we need to reconsider the position and functionality of the option layer, cause each individual layer has its design and cost. 

BTW. In general, the interface parameters include three parts, the configurations(str, int, bool..), the feature flags(bools..) and the functional structures(IO, client, etc). The functional structures are injected into the lower layer by the middle layer, and obtained by the factory provided by the lower layer. We can design and implement APIs more standardly.

**Comment by elliotxx** [comment]
@ldxdl OK, the interface design will continue to be optimized, and we will submit PR separately to solve.

**Review by ldxdl — APPROVED** [review]
LGTM

**Pull request merged**